internal/common/testing/container/container.go

```diff
@@ -0,0 +1,175 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package container
+
+import (
+	"bufio"
+	"context"
+	"net"
+	"testing"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/go-connections/nat"
+	"github.com/stretchr/testify/require"
+)
+
+type ContainerID string
+
+type Containers struct {
+	cli               client.APIClient
+	runningContainers []ContainerID
+	t                 *testing.T
+}
+
+type Container struct {
+	ID           ContainerID
+	waitForPorts []uint16
+	ports        nat.PortMap
+}
+
+func (c *Container) AddrForPort(port uint16) string {
+	for dockerPort, bindings := range c.ports {
+		if dockerPort.Int() == int(port) && len(bindings) > 0 {
+			return net.JoinHostPort(bindings[0].HostIP, bindings[0].HostPort)
+		}
+	}
+
+	return ""
+}
+
+type ContainerOption func(c *Container)
+
+// WithPortReady declares that the provided container port should be able to be connected
+// to before returning from StartContainer.
+func WithPortReady(port uint16) ContainerOption {
+	return func(c *Container) {
+		c.waitForPorts = append(c.waitForPorts, port)
+	}
+}
+
+// New wraps the provided testing.T instance so that any containers started will be cleaned
+// up when the test completes.
+func New(t *testing.T) *Containers {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		t.Fatalf("failed creating docker client from environment: %v", err)
+	}
+
+	ctx := &Containers{cli: cli, t: t}
+
+	t.Cleanup(func() {
+		ctx.Cleanup()
+	})
+
+	return ctx
+}
+
+// Cleanup stops and removes all containers. It is called automatically when the test completes
+// if called from New() but can also be called manually if needed. It is idempotent.
+func (c *Containers) Cleanup() {
+	for _, con := range c.runningContainers {
+		if err := c.cli.ContainerRemove(context.Background(), string(con), types.ContainerRemoveOptions{
+			RemoveVolumes: true,
+			Force:         true,
+		}); err != nil {
+			c.t.Logf("failed removing container %v: %v", con, err)
+		}
+	}
+	c.runningContainers = nil
+	c.cli.Close()
+}
+
+// StartContainer starts a container with the given image and zero or more ContainerOptions.
+func (c *Containers) StartContainer(image string, opts ...ContainerOption) Container {
```

Comment:
Might also be useful to be able to pass in path to a Dockerfile from which the image can be built.

Reply:
Yep we'll need that eventually.